### PR TITLE
Fix line_dash list ambiguity for all glyph types

### LINKED ISSUE #13838: [BUG] Using array/list in line_dash for hspan doesn't work
### Software versions

Python version        :  3.12.2 (main, Feb 21 2024, 00:00:00) [GCC 14.0.1 20240217 (Red Hat 14.0.1-0)]
IPython version       :  (not installed)
Tornado version       :  6.4
Bokeh version         :  3.4.1
BokehJS static path   :  /var/home/michael/src/koding/python/bokeh_hspan/.venv/lib/python3.12/site-packages/bokeh/server/static
node.js version       :  (not installed)
npm version           :  (not installed)
jupyter_bokeh version :  (not installed)
Operating system      :  Linux-6.8.7-300.fc40.x86_64-x86_64-with-glibc2.39

### Browser name and version

Firefox 125.0.2

### Jupyter notebook / Jupyter Lab version

_No response_

### Expected behavior

Specifying line_dash in hspan should produce an hspan line with the given dash pattern.

### Observed behavior

Specifying line_dash as an array/list causes the plot to not display properly. I can't see the hspan line, and when I pan the plot it starts to slide down towards the lower right corner (see attached screencast).

Using string names such as "solid" and "dashed" works ok.

### Example code

```Python
from bokeh.plotting import figure, show

plot = figure()
plot.line(x=[-5, 5], y=[-5, 5])
plot.hspan(y=0, line_dash=[5, 5])

show(plot)
```

### Stack traceback or browser console output

_No response_

### Screenshots

[Screencast from 2024-04-25 15-17-45.webm](https://github.com/bokeh/bokeh/assets/33153877/0056a438-9e76-41c3-9eeb-8dd525f5d7c0)

### PR DESCRIPTION
**Pull request by canavandl**
## Summary
Fixes #13838 - Resolves the issue where `line_dash=[6, 3]` was incorrectly treated as per-glyph data instead of a scalar dash pattern.

### Core Fix (`src/bokeh/plotting/_renderer.py`)
- Added handling for both `DashPattern` (Line glyphs) and `DashPatternSpec` (Patches glyphs)
- Placed check BEFORE "non-dataspecs" early exit (critical for Line glyphs)
- Converts NumPy arrays to lists for proper BokehJS serialization
- Added clarifying comments distinguishing DashPattern vs DashPatternSpec

### Test Coverage (`tests/unit/bokeh/plotting/test__renderer.py`)
- Added 15 comprehensive tests in `Test__process_sequence_literals` class
- Tests multiple input types: lists, tuples, NumPy arrays (various dtypes)
- Tests multiple glyph types: Line, Patches, HSpan, MultiLine
- Tests edge cases: empty patterns, float arrays (validation failure)
- Includes regression test for per-glyph dash patterns via ColumnDataSource

### Now Works ✅
```python
# Lists/tuples as scalar patterns
p.line(x, y, line_dash=[6, 3])
p.hspan(y=0, line_dash=[5, 5])  # Original issue #13838

# NumPy arrays as scalar patterns  
p.line(x, y, line_dash=np.array([6, 3]))
p.patches(xs, ys, line_dash=np.array([10, 5, 2, 5]))

# Complex patterns
p.line(x, y, line_dash=[10, 5, 2, 5])
```

### Still Works ✅
```python
# String patterns (unchanged)
p.line(x, y, line_dash="dashed")

# Per-glyph patterns via ColumnDataSource (regression tested)
source = ColumnDataSource(data={
    'xs': [[0, 1], [1, 2]],
    'ys': [[0, 1], [1, 2]],
    'line_dash': [[6, 3], [10, 5]]
})
p.multi_line('xs', 'ys', line_dash='line_dash', source=source)
```

## Implementation Details

The fix uses `allspecs = glyphclass.properties(_with_props=True)` to get ALL properties (not just DataSpecs), then checks:
- `isinstance(allspecs[var], (DashPatternSpec, DashPattern))`
- For NumPy arrays: validates `val.ndim == 1 and val.dtype.kind in ('i', 'u')`  
- Converts via `kwargs[var] = val.tolist()`

## Test Results
All 15 new tests pass ✅
- 8 existing `test__renderer.py` tests still pass
- No regressions detected

### CODE REVIEW

File: tests/unit/bokeh/plotting/test__renderer.py [hunk 1]
```diff
@@ -108,6 +117,155 @@ def test_user_specified_mute_glyph(self) -> None:
...
+        assert "line_dash" not in r.data_source.data
+
+    def test_line_dash_as_tuple(self) -> None:
+        """line_dash=(6, 3) should be treated as a scalar dash pattern"""
+        p = figure()
+        r = p.line([1, 2, 3], [1, 2, 3], line_dash=(6, 3))
+
+        assert isinstance(r.glyph, Line)
+        # Tuples get converted to lists internally
+        assert r.glyph.line_dash == [6, 3] or r.glyph.line_dash == (6, 3)
```
Comment (bryevdv) [inline tests/unit/bokeh/plotting/test__renderer.py]: We can't nail this down to one outcome?

File: tests/unit/bokeh/plotting/test__renderer.py [hunk 2]
```diff
@@ -108,6 +117,155 @@ def test_user_specified_mute_glyph(self) -> None:
...
+        assert "line_dash" not in r.data_source.data
+
+    def test_line_dash_numpy_uint8_array(self) -> None:
+        """line_dash with numpy uint8 array should work"""
+        p = figure()
+        r = p.line([1, 2, 3], [1, 2, 3], line_dash=np.array([6, 3], dtype=np.uint8))
+
+        assert isinstance(r.glyph, Line)
+        assert list(r.glyph.line_dash) == [6, 3]
+        assert "line_dash" not in r.data_source.data
```
Comment (bryevdv) [inline tests/unit/bokeh/plotting/test__renderer.py]: prefer if all the numpy cases were collapse to one test function using `pytest.mark.parametrize` for the dtype

File: tests/unit/bokeh/plotting/test__renderer.py [hunk 3]
```diff
@@ -108,6 +117,155 @@ def test_user_specified_mute_glyph(self) -> None:
...
+
+    def test_line_dash_float_list_fails(self) -> None:
+        """line_dash with float list should fail validation"""
+        p = figure()
+        with pytest.raises(ValueError, match="failed to validate"):
+            p.line([1, 2, 3], [1, 2, 3], line_dash=[6.0, 3.0])
```
Comment (bryevdv) [inline tests/unit/bokeh/plotting/test__renderer.py:213]: should probably also test the the case with some but not all values being floats

File: tests/unit/bokeh/plotting/test__renderer.py [hunk 4]
```diff
@@ -108,6 +117,155 @@ def test_user_specified_mute_glyph(self) -> None:
...
+            'line_dash': [[6, 3], [10, 5], [2, 4]],  # Per-glyph patterns
+        })
+        r = p.multi_line('xs', 'ys', line_dash='line_dash', source=source)
+
+        assert isinstance(r.glyph, MultiLine)
+        # line_dash should reference the field name
+        assert r.glyph.line_dash == 'line_dash'
+        # Data should be in the source
+        assert 'line_dash' in r.data_source.data
+        assert r.data_source.data['line_dash'] == [[6, 3], [10, 5], [2, 4]]
```
Comment (bryevdv) [inline tests/unit/bokeh/plotting/test__renderer.py:260]: Do we also care to support passing this case like this for "multi" glyphs?
```python
p.multi_line(..., line_dash=[[6, 3], [10, 5], [2, 4]])
```

File: src/bokeh/plotting/_renderer.py [hunk 5]
```diff
@@ -261,6 +261,7 @@ def _pop_renderer_args(kwargs: Attrs) -> Attrs:
 def _process_sequence_literals(glyphclass: type[Glyph], kwargs: Attrs, source: ColumnarDataSource, is_user_source: bool) -> list[str]:
     incompatible_literal_spec_values: list[str] = []
     dataspecs = glyphclass.dataspecs()
+    allspecs = glyphclass.properties(_with_props=True)
```
Comment (bryevdv) [inline src/bokeh/plotting/_renderer.py]: Could this use a better name? "allspecs" suggests to me that these would all be dataspec properties but that is not the case, no?

File: tests/unit/bokeh/plotting/test__renderer.py (hunk 4 above)
Comment (bryevdv) [reply tests/unit/bokeh/plotting/test__renderer.py:260]: I guess passing that way would require _not_ using a user-created data source as well, maybe that's uncommon enough to not worry about.

File: src/bokeh/plotting/_renderer.py
Comment (bryevdv) [inline src/bokeh/plotting/_renderer.py:314]: I wonder can the order of these conditionals be swapped? Above we type check first for dash-y properties first then in side those blocks handle numpy arrays, and it seems a bit odd to do things in the opposite order elsewhere.

**Comment by bryevdv** [comment]
The code in `_process_sequence_literals` (mostly mine AFAIK) was already pretty hilariously convoluted. I wonder what ways it could be cleaned up. I guess ideally the properties/dataspecs would handle more of this logic but I suppose there are/were reasons that was not the case. Anyway, jsut a few small comments otherwise LGTM

File: tests/unit/bokeh/plotting/test__renderer.py (hunk 1 above)
Comment (canavandl) [reply tests/unit/bokeh/plotting/test__renderer.py]: fixed

File: tests/unit/bokeh/plotting/test__renderer.py (hunk 2 above)
Comment (canavandl) [reply tests/unit/bokeh/plotting/test__renderer.py]: done

File: tests/unit/bokeh/plotting/test__renderer.py (hunk 3 above)
Comment (canavandl) [reply tests/unit/bokeh/plotting/test__renderer.py:213]: done

File: tests/unit/bokeh/plotting/test__renderer.py (hunk 4 above)
Comment (canavandl) [reply tests/unit/bokeh/plotting/test__renderer.py:260]: It works automatically. I added a test to confirm and a test to confirm there's a `warn` when the array length is wrong

File: src/bokeh/plotting/_renderer.py (hunk 5 above)
Comment (canavandl) [reply src/bokeh/plotting/_renderer.py]: renamed to `all_properties`

File: src/bokeh/plotting/_renderer.py
Comment (canavandl) [reply src/bokeh/plotting/_renderer.py:314]: done

**Comment by canavandl** [comment]
>The code in _process_sequence_literals (mostly mine AFAIK) was already pretty hilariously convoluted. I wonder what ways it could be cleaned up. I guess ideally the properties/dataspecs would handle more of this logic but I suppose there are/were reasons that was not the case. 

refactored

**Review by bryevdv — APPROVED** [review]

**Comment by mosc9575** [comment]
I think this should be added to `3.9.1` - see #15089.
